src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/incentives/IncentivesService.kt

```diff
@@ -85,6 +89,89 @@ class IncentivesService(
       } ?: throw NotFoundException("Current Incentive not found, booking id $bookingId")
   }
 
+  fun createGlobalIncentiveLevel(createIncentiveRequest: ReferenceCode): ReferenceCode {
+    return incentiveReferenceCodeRepository.findByIdOrNull(IEPLevel.pk(createIncentiveRequest.code))
+      ?.let { ReferenceCode(it.code, it.domain, it.description, it.active) }
+      .also { log.info("Global IEP level: $createIncentiveRequest already exists") } // TODO call update??
+      ?: let {
+        incentiveReferenceCodeRepository.save(
+          IEPLevel(
+            createIncentiveRequest.code,
+            createIncentiveRequest.description,
+            createIncentiveRequest.active,
+          ),
+        ).let { ReferenceCode(it.code, it.domain, it.description, it.active) }.also {
+          telemetryClient.trackEvent(
+            "global-incentive-level-created",
+            mapOf(
+              "code" to it.code,
+              "active" to it.active.toString(),
+              "description" to it.description,
+            ),
+            null,
+          )
+        }
+      }
+  }
+
+  fun updateGlobalIncentiveLevel(code: String, updateIncentiveRequest: UpdateGlobalIncentiveRequest): ReferenceCode {
+    return incentiveReferenceCodeRepository.findByIdOrNull(IEPLevel.pk(code))
+      ?.let {
+        it.active = updateIncentiveRequest.active
+        it.description = updateIncentiveRequest.description
+        telemetryClient.trackEvent(
+          "global-incentive-level-updated",
+          mapOf(
+            "code" to code,
+            "active" to it.active.toString(),
+            "description" to it.description,
+          ),
+          null,
+        )
+        return ReferenceCode(it.code, it.domain, it.description, it.active)
+      } ?: throw NotFoundException("Incentive level: $code not found")
+  }
+
+  fun getGlobalIncentiveLevel(code: String): IEPLevel {
+    return incentiveReferenceCodeRepository.findByIdOrNull(IEPLevel.pk(code))
+      ?: throw NotFoundException("Incentive level: $code not found")
+  }
+
+  fun deleteGlobalIncentiveLevel(code: String) {
+    incentiveReferenceCodeRepository.findByIdOrNull(IEPLevel.pk(code))
+      ?.let {
+        incentiveReferenceCodeRepository.delete(it)
+        telemetryClient.trackEvent(
+          "global-incentive-level-deleted",
+          mapOf(
+            "code" to code,
+          ),
+          null,
+        )
+        log.info("Global Incentive level deleted: $code")
+      } ?: log.info("Global Incentive level deletion request for: $code ignored. Level does not exist")
+  }
+
+  fun reorderGlobalIncentiveLevels(codeList: List<String>) {
+    codeList.mapIndexed { index, levelCode ->
```

Comment:
I wonder if we need to handle if a list is provided that is missing a level that exists in NOMIS (probably will never happen except in test environments)?

Reply:
yes - there would be 2 with same sequence potentially (which is allowed) 
I could add a general count check at the beginning and log an error if there is a mismatch between the number provided and those in NOMIS